dag_in_context/src/interval_analysis.egg

```diff
@@ -249,4 +265,73 @@
        (= (IntB hi) (hi-bound v))
       )
       ((set (hi-bound lhs) (IntB  hi)))
-      :ruleset interval-analysis)
\ No newline at end of file
+      :ruleset interval-analysis)
+
+
+
+; If the If takes a tuple
+(rule (
+        ; expr < value
+        (= pred (Bop (LessThan) expr value))
+        (= if_e (If pred inputs then else))
+        ; the left operand of the < is an input to the if region
+        (= expr (Get inputs i))
+        ; the right operand of the < has an upper bound
+        (= (IntB v) (hi-bound value))
+        ; context node inside the if region
+        (= ctx (InContext (InIf true pred inputs) (Arg ty)))
+        (HasType inputs ty)
+        ; usage of the value x inside the if region
+        (Get ctx i)
+      )
+      ((set (hi-bound (Get ctx i)) (IntB (- v 1)))) ; expr < value was true, so we know expr < (hi-bound value) - 1
```

Comment:
One thing I'm concerned about with the current interval analysis is integer overflow
What happens when this overflows? I think it makes the interval analysis unsound
A simple fix would be to add to the query that v is greater than INT_MIN

Reply:
All the other interval rules probably have this problem, so lets not solve it in this PR.
We don't necessarily need to solve it unless we find it happens in benchmarks

Reply:
Will handle as follow-up PR